streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1791,6 +1794,52 @@ protected int processStreamThread(final Consumer<StreamThread> consumer) {
         return copy.size();
     }
 
+    /**
+     * Returns the internal clients' assigned {@code client instance ids}.
+     *
+     * @return the internal clients' assigned instance ids used for metrics collection.
+     *
+     * @throws IllegalStateException If {@code KafkaStreams} is not running.
+     * @throws TimeoutException Indicates that a request timed out.
+     */
+    public ClientInstanceIds clientInstanceIds(final Duration timeout) {
+        if (state().hasNotStarted()) {
+            throw new IllegalStateException("KafkaStreams has not been started, you can retry after calling start().");
+        }
+        if (state().isShuttingDown() || state.hasCompletedShutdown()) {
+            throw new IllegalStateException("KafkaStreams has been stopped (" + state + ").");
+        }
+
+        final ClientInstanceIdsImpl clientInstanceIds = new ClientInstanceIdsImpl();
+
+        KafkaFuture<Uuid> globalThreadFuture = null;
+        if (globalStreamThread != null) {
+            globalThreadFuture = globalStreamThread.globalConsumerInstanceId(timeout);
```

Comment:
You are applying the same Duration to all of the calls, but the caller would expect the cumulative duration not to exceed the parameter supplied.

Reply:
How strict we can obey the given `timeout` is somewhat tricky, given that we need to call `clientInstanceId()` for each client we have. -- The idea was to basically "fan-out" all these calls and to them in parallel (note that `globalConsumerInstanceId` will return immediately and not block, but hand the execution from the user thread to the `GlobalStreamThread`; that's why we return a Future) -- thus it should be ok to provide the same timeout to each call (as all of them are done in parallel)?

If you have any good suggestion how it could be done better, let me know.

Reply:
OK, it sounds like you have a plan.

Reply:
I like to think so :)